### DISCUSSION TITLE
Do I have to enable CORS to use Supabase Storage with S3?

### DISCUSSION THREAD

**Discussion by ColeBlender** [supabase/supabase#23198] [Questions]
I have been using AWS S3 and want to switch to Supabase Storage. I already have all of my code written using the S3 SDK so I want to keep using that. I've followed the guide here [https://supabase.com/docs/guides/storage/s3/authentication](url) but it's still not working. Here is my code:
This is my route handler to generate a signed url

`
import { protect } from "@/lib/auth";
import { toS3SafeFormat } from "@/lib/utils";
import { S3Client, PutObjectCommand } from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
import { NextRequest, NextResponse } from "next/server";

// upload-image
export async function POST(req: NextRequest) {
  try {
    await protect();

    const { fileName, folder } = await req.json();

    if (!folder || !fileName) {
      return NextResponse.json(
        { error: "Proper image data not provided" },
        { status: 400 },
      );
    }

    const fileKey = `${folder}/${Date.now().toString()}-${toS3SafeFormat(fileName)}`;

    const s3: any = new S3Client({
      forcePathStyle: true,
      region: "us-west-1",
      endpoint: `${process.env.NEXT_PUBLIC_SUPABASE_URL}/storage/v1/s3`,
      credentials: {
        accessKeyId: process.env.AWS_ACCESS_KEY_ID!,
        secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY!,
      },
    });

    const command: any = new PutObjectCommand({
      Bucket: process.env.NEXT_PUBLIC_AWS_BUCKET_NAME,
      Key: fileKey,
    });

    const url = await getSignedUrl(s3, command, { expiresIn: 60 });

    return NextResponse.json(
      {
        url,
        fileKey,
        errorMessage: null,
      },
      { status: 200 },
    );
  } catch (error) {
    return NextResponse.json(
      { errorMessage: "Failed to generate upload url" },
      { status: 500 },
    );
  }
}

`

This is my code to upload it which takes place on the client

`
import { getErrorMessage } from "./utils";
import imageCompression from "browser-image-compression";

export async function uploadToS3(imageFile: File, folder: string) {
  try {
    const { url, fileKey, errorMessage } = await fetch(
      "/api/s3/get-upload-url",
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          fileName: imageFile.name,
          folder,
        }),
      },
    ).then((res) => res.json());

    if (errorMessage) {
      throw new Error(errorMessage);
    }

    try {
      imageFile = await imageCompression(imageFile, {
        maxSizeMB: 1,
      });
    } catch (error) {
      throw new Error("Image compression failed");
    }

    const uploadResponse = await fetch(url, {
      method: "PUT",
      body: imageFile,
    });

    console.log(uploadResponse);

    if (!uploadResponse.ok) {
      throw new Error("Image upload failed");
    }

    const s3ImageUrl = `${process.env.NEXT_PUBLIC_SUPABASE_URL}/storage/v1/object/public/${process.env.NEXT_PUBLIC_AWS_BUCKET_NAME}/${fileKey}`;

    return { s3ImageUrl, errorMessage: null };
  } catch (error) {
    return { errorMessage: getErrorMessage(error) };
  }
}
`

**Comment by bombillazo** [comment]
I am also interested in this.

**Comment by vignesh1398-vz** [comment]
Any update on this issue?

**Comment by algi** [comment]
Same issue, apart from that I'm using Rails with ActionText that generates those signed URLs from the server.

**Comment by rnons** [comment]
A workaround is including `_` in your bucket name. Some observations:

```js
const url = await getSignedUrl(
  s3Client,
  new PutObjectCommand({
    Bucket: 'upload',
    Key: 'dummy.pdf',
  }),
);
```
returns `https://upload.<project-id>.storage.supabase.co/storage/v1/s3/dummy.pdf?***`, which triggers CORS error



```js
const url = await getSignedUrl(
  s3Client,
  new PutObjectCommand({
    Bucket: 'upload_pending',
    Key: 'dummy.pdf',
  }),
);
```

returns `https://<project-id>.storage.supabase.co/storage/v1/s3/upload_pending/dummy.pdf?***`, which works fine without CORS error


In the first case bucket name becomes a subdomain. In the second case, bucket name is part of the pathname.
